src/VirtoCommerce.CatalogPersonalizationModule.Web/ExportImport/PersonalizationExportImport.cs

```diff
@@ -1,60 +1,94 @@
 using VirtoCommerce.Platform.Core.Common;
 using VirtoCommerce.Platform.Core.ExportImport;
+using VirtoCommerce.Platform.Data.ExportImport;
 
 namespace VirtoCommerce.CatalogPersonalizationModule.Web.ExportImport
 {
-    public sealed class BackupObject
-    {
-        public ICollection<TaggedItem> TaggedItems { get; set; }
-    }
-
     public sealed class PersonalizationExportImport
     {
         private readonly ITaggedItemService _taggedItemService;
         private readonly ITaggedItemSearchService _taggedItemSearchService;
+        private readonly JsonSerializer _serializer;
+        private readonly int _batchSize = 50;
 
-        public PersonalizationExportImport(ITaggedItemService taggedItemService, ITaggedItemSearchService taggedItemSearchService)
+        public PersonalizationExportImport(ITaggedItemService taggedItemService, ITaggedItemSearchService taggedItemSearchService, JsonSerializer jsonSerializer)
         {
             _taggedItemService = taggedItemService;
             _taggedItemSearchService = taggedItemSearchService;
+            _serializer = jsonSerializer;
         }
 
-        public async Task DoExportAsync(Stream backupStream, Action<ExportImportProgressInfo> progressCallback)
+        public async Task DoExportAsync(Stream outStream, Action<ExportImportProgressInfo> progressCallback, ICancellationToken cancellationToken)
         {
-            var backupObject = await GetBackupObjectAsync(progressCallback);
-            backupObject.SerializeJson(backupStream);
-        }
+            cancellationToken.ThrowIfCancellationRequested();
+            var progressInfo = new ExportImportProgressInfo {Description = "loading data..."};
+            progressCallback(progressInfo);
 
-        public async Task DoImportAsync(Stream backupStream, Action<ExportImportProgressInfo> progressCallback)
-        {
-            var backupObject = backupStream.DeserializeJson<BackupObject>();
-            var progressInfo = new ExportImportProgressInfo();
+            using (var sw = new StreamWriter(outStream, System.Text.Encoding.UTF8))
+            using (var writer = new JsonTextWriter(sw))
+            {
+                await writer.WriteStartObjectAsync();
 
-            progressInfo.Description = String.Format("{0} tagged items importing...", backupObject.TaggedItems.Count);
-            progressCallback(progressInfo);
+                var taggedItems = await _taggedItemSearchService.SearchTaggedItemsAsync(new TaggedItemSearchCriteria {Take = int.MaxValue});
+                progressInfo.Description = $"{taggedItems.TotalCount} tagged items loading...";
+                progressCallback(progressInfo);
 
-            await _taggedItemService.SaveChangesAsync(backupObject.TaggedItems.ToArray());
+                await writer.WritePropertyNameAsync("TaggedItemsTotalCount");
+                await writer.WriteValueAsync(taggedItems.TotalCount);
+
+                await writer.WritePropertyNameAsync("TaggedItems");
```

Comment:
Could you use paged generic methods like here, to not read all data at one query?
https://github.com/VirtoCommerce/vc-module-catalog/blob/dev/src/VirtoCommerce.CatalogModule.Data/ExportImport/CatalogExportImport.cs#L71

Reply:
updated